chore: add example for how to use TrackConsumersPool

wiedld opened the pull request:
## Which issue does this PR close?

- Part of https://github.com/apache/datafusion/issues/16904


## Rationale for this change

Expand docs, with code examples, for the current functionality for `TrackConsumersPool`.

Note that this functionality will be expanded after https://github.com/apache/datafusion/pull/17021, after which we should add a doc section for using the realtime tracking of `ConsumerMemoryMetrics`.

## What changes are included in this PR?

Docs for the existing functionality.

## Are these changes tested?

yes

## Are there any user-facing changes?

Docs.

wiedld commented on datafusion/execution/src/memory_pool/pool.rs:
````diff
@@ -295,9 +295,228 @@ impl TrackedConsumer {
+/// # }
+/// ```
+///
+/// # How to use in ExecutionPlan implementations
+///
+/// When implementing custom ExecutionPlans that need to use significant memory, you should
+/// use the memory pool to track and limit memory usage:
````
I wasn't sure about this section. It's explaining how the `MemoryReservation` should be used; which applies to any `MemoryPool`.

It's just that our specific memory pool implementation, `TrackConsumersPool`, it will be aggregating the reservations per memory consumer. So it felt like added this part helped draw the pieces together for other contributors, maybe?

alamb replied:
I think this is good content but as you point out, will be hard to find here. 

I recommend you move it to `ExecutionPlan` where people are more likely to find it, perhaps similarly to the code here:

https://docs.rs/datafusion/latest/datafusion/physical_plan/trait.ExecutionPlan.html#implementation-examples

2010YOUY01 reviewed (approved):
Thank you. 

I have read through it and I think it's a great explanation. What do you think about moving them to `datafusion-examples` and referencing the example here? Potential users might want to experiment by modifying it, since this example demonstrates many functionalities.

alamb commented on datafusion/execution/src/memory_pool/pool.rs:
```diff
@@ -311,6 +531,36 @@ pub struct TrackConsumersPool<I> {
+    /// * `top` - The number of top memory consumers to include in error messages
+    ///
+    /// # Note
+    /// In most cases, you should use [`RuntimeEnvBuilder::with_memory_limit()`](crate::runtime_env::RuntimeEnvBuilder::with_memory_limit)
```
👍

alamb commented on datafusion/execution/src/memory_pool/pool.rs:
````diff
@@ -295,9 +295,228 @@ impl TrackedConsumer {
+/// # }
+/// ```
+///
+/// # Runtime Monitoring
+///
+/// TODO(kosiew/wiedld):  Docs to be added after <https://github.com/apache/datafusion/pull/17021/>.
````
I think this is also a good thing, but a larger topic than just `TrackedPool` -- maybe we could move some of this to an example and leave a link here. I will comment more elsewhere

alamb reviewed (approved):
Thank you @wiedld  and @2010YOUY01 

This is a very nice improvement to the overall documentation, as memory management I think is a core feature of datafusion and not as well documented / exampled as other areas yet. 

I think this PR is a step forward  and could be merged as is, but I have a few other suggestions. 

The examples are neat, but it is unfortunate that we can't actually run them because `datafusion-execution` can't rely on the `datafusion` crate

What would you think about moving some/all of the code to a real example in https://github.com/apache/datafusion/tree/main/datafusion-examples/examples

Something like `memory_pools.rs`

We could then have several functions in that example that show the various APIs, similar to https://github.com/apache/datafusion/blob/1d4d74bedafb73138237d8194d36778373d39a8e/datafusion-examples/examples/expr_api.rs#L50-L57

So functions corresponding to your headings:
1. Automatic tracking via RuntimeEnvBuilder
2. ExecutionPlan implementations
3. Runtime Monitoring
4. (future) The new API for detailed memory analysis

We could then leave links to the example from ExecutionPlan and TrackedPool?

wiedld commented:
Thank you for the reviews @2010YOUY01 and @alamb .

I moved the examples to `datafusion-examples`, and fleshed them out quite a bit. I also added references to these examples in the TrackConsumersPool and ExecutionPlan docs. Please let me know if any of these changes are a bit too much.

I didn't yet add a section for `Runtime Monitoring` as the report functionality is really about producing a formatted string for error messages. I'm hoping that the followup work associated with both CLI memory tracking flag, and potential API changes from issue https://github.com/apache/datafusion/issues/16904, will give us a more useful monitoring tool.

wiedld commented on datafusion-examples/examples/memory_pool_tracking.rs:
```diff
@@ -0,0 +1,191 @@
+use std::sync::Arc;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
```
Output is:

```
=== DataFusion Memory Pool Tracking Example ===

Example 1: Automatic Usage with RuntimeEnvBuilder
------------------------------------------------
Success case: Normal operation with sufficient memory
✓ Created table with memory tracking enabled
✓ Query executed successfully. Found 1 rows

--------------------------------------------------
Error case: Triggering memory limit error with detailed error messages
✓ Expected memory limit error during data processing:
Error: Not enough memory to continue external sort. Consider increasing the memory limit, or decreasing sort_spill_reservation_bytes
caused by
Resources exhausted: Additional allocation failed with top memory consumers (across reservations) as:
  ExternalSorter[12]#152(can spill: true) consumed 0.0 B,
  ExternalSorter[9]#143(can spill: true) consumed 0.0 B,
  HashJoinInput[7]#136(can spill: false) consumed 0.0 B,
  RepartitionExec[0]#94(can spill: false) consumed 0.0 B,
  RepartitionExec[7]#126(can spill: false) consumed 0.0 B.
Error: Failed to allocate additional 10.0 MB for ExternalSorterMerge[0] with 0.0 B already allocated for this reservation - 4.8 MB remain available for the total pool

Note: The error message above shows which memory consumers
were using the most memory when the limit was exceeded.

============================================================

Example 2: Manual Memory Consumer Tracking
------------------------------------------
Created TrackConsumersPool with 1KB limit, tracking top 3 consumers
BigOperation: reserved 400 bytes
MediumOperation: reserved 200 bytes
SmallOperation: reserved 100 bytes
Total reserved: 700 bytes out of 1000 byte limit

Attempting to reserve 500 bytes (would exceed limit)...
Expected failure with detailed error message:
Resources exhausted: Additional allocation failed with top memory consumers (across reservations) as:
  BigOperation#157(can spill: false) consumed 400.0 B,
  MediumOperation#158(can spill: false) consumed 200.0 B,
  SmallOperation#159(can spill: false) consumed 100.0 B.
Error: Failed to allocate additional 500.0 B for FailingOperation with 0.0 B already allocated for this reservation - 300.0 B remain available for the total pool

============================================================
```

wiedld commented on datafusion-examples/examples/memory_pool_execution_plan.rs:
```diff
@@ -0,0 +1,324 @@
+use std::sync::Arc;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
```
Output is:

```
=== DataFusion ExecutionPlan Memory Tracking Example ===

Example: Custom Memory-Aware BufferingExecutionPlan
---------------------------------------------------
Executing BufferingExecutionPlan with memory tracking...
Memory limit: 15000 bytes - should trigger spill on later batches

Added batch of 10776 bytes, total buffered: 10776 bytes
Memory limit reached, spilling 10776 bytes to disk
Spill #1: Writing 10776 bytes to disk
Added batch of 10776 bytes, total buffered: 10776 bytes
Memory limit reached, spilling 10776 bytes to disk
Spill #2: Writing 10776 bytes to disk
Added batch of 14872 bytes, total buffered: 14872 bytes
Memory limit reached, spilling 14872 bytes to disk
Spill #3: Writing 14872 bytes to disk
Added batch of 10776 bytes, total buffered: 10776 bytes
Finished processing, released 10776 bytes

Successfully executed BufferingExecutionPlan!

The BufferingExecutionPlan processed 4 input batches and
demonstrated memory tracking with spilling behavior when the
memory limit was exceeded by later batches.
Check the console output above to see the spill messages.
```

alamb commented:
This looks good -- I plan a more detailed review shortly

alamb commented on datafusion/execution/src/memory_pool/pool.rs:
```diff
@@ -295,9 +295,72 @@ impl TrackedConsumer {
+///
+/// # Automatic Usage with RuntimeEnvBuilder
+///
+/// The easiest way to use `TrackConsumersPool` is through [`RuntimeEnvBuilder::with_memory_limit()`](crate::runtime_env::RuntimeEnvBuilder::with_memory_limit),
```
this example seems redundant with the example you added, so I'll remove it for now

alamb commented on datafusion-examples/examples/memory_pool_tracking.rs:
```diff
@@ -0,0 +1,191 @@
+
+/// Example 2: Manual tracking with custom consumers
+///
+/// This shows how to manually create and use memory consumers with tracking.
```
I am not sure this example adds much value as I don't expect people to generate their own consumers outside the context of writing an operator (which is covered by the other example)

Let's remove it from this PR for now

alamb commented on datafusion-examples/examples/memory_pool_tracking.rs:
```diff
@@ -0,0 +1,191 @@
+    // Error case: Create a runtime with low memory limit to trigger errors
+    println!("Error case: Triggering memory limit error with detailed error messages");
+
+    // Use a WITH query that generates data and then processes it to trigger memory usage
```
I think this will be a nice base to start from when implementing more sophisticated memory reporting / tracking

alamb commented on datafusion-examples/examples/memory_pool_execution_plan.rs:
```diff
@@ -0,0 +1,324 @@
+
+    // Create smaller batches to ensure we get multiple RecordBatches from the scan
+    // Make each batch smaller than the memory limit to force multiple batches
+    let batch1 = RecordBatch::try_new(
```
maybe this would be a good use for the https://docs.rs/arrow/latest/arrow/array/macro.record_batch.html macro

alamb commented on datafusion-examples/examples/memory_pool_execution_plan.rs:
```diff
@@ -0,0 +1,324 @@
+    }
+
+    fn spill_to_disk(&mut self) -> Result<()> {
+        // Simulate writing buffer to disk
```
"simulate" lol

alamb reviewed (approved):
Thank you @wiedld and @2010YOUY01 

I took the liberty of pushing some commits to make the documentation more concise and the examples shorter.
